### PR TITLE
Add TypeFlags for TyKind in chalk-ir

### LINKED ISSUE #627: add flags to `InternedTy` and `InternedRegion` types
In rustc, we have a "flags" system that propagates information about a type upward and which can be used to shortcircuit a lot of work. The [`TypeFlags`] propagates various bits. Each time a type is interned, the flags from any subtypes are combined to form the flags for the new interned type. This allows us to do things like skip subtrees that don't contain the sort of thing we are looking for efficiently.

In the meeting today we discussed different ways of doing this. I'll add a comment with more notes.

### PR DESCRIPTION
**Pull request by chfont**
Closes #627

### CODE REVIEW

File: chalk-ir/src/lib.rs [hunk 1]
```diff
@@ -327,11 +328,148 @@ pub struct Ty<I: Interner> {
...
+        TyKind::Adt(_, substitution)
+        | TyKind::AssociatedType(_, substitution)
+        | TyKind::Tuple(_, substitution)
+        | TyKind::Closure(_, substitution)
+        | TyKind::Generator(_, substitution)
+        | TyKind::GeneratorWitness(_, substitution)
+        | TyKind::FnDef(_, substitution) => compute_substitution_flags(substitution, interner),
+        TyKind::Scalar(_) | TyKind::Str | TyKind::Never | TyKind::Foreign(_) => TypeFlags::empty(),
+        TyKind::OpaqueType(_, substitution) => {
+            TypeFlags::HAS_TY_OPAQUE | compute_substitution_flags(substitution, interner)
```
Comment (matthewjasper) [inline chalk-ir/src/lib.rs:410]: So, looking through how this is used in rustc  I think that the majority of code would not expect this it to be set for `TyKind::OpaqueType`, only `AliasTy::Opaque`.

File: tests/test/type_flags.rs [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+                    value: chalk_ir::ConstValue::InferenceVar(chalk_ir::InferenceVar::from(0)),
+                },
+            )),
+        )];
+    let opaque_ty_kind = TyKind::OpaqueType(
+        chalk_ir::OpaqueTyId {
+            0: chalk_integration::interner::RawId { index: 0 },
+        },
+        chalk_ir::Substitution::from_iter(&chalk_integration::interner::ChalkIr, x),
+    );
```
Comment (jackh726) [inline tests/test/type_flags.rs:39]: As above, can we use the `ty` macro?

**Review by jackh726 — CHANGES_REQUESTED** [review]
My comments can apply to the rest of the tests I didn't comment on

File: tests/test/type_flags.rs [hunk 3]
```diff
@@ -0,0 +1,152 @@
+use chalk_integration::{empty_substitution, lifetime, ty};
```
[the remark this answers is not in the document]
Comment (jackh726) [reply tests/test/type_flags.rs:2]: This can be moved later.

File: tests/test/type_flags.rs [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+                &chalk_integration::interner::ChalkIr,
+                chalk_ir::ConstData {
+                    ty: Ty::new(
+                        &chalk_integration::interner::ChalkIr,
+                        TyKind::Placeholder(PlaceholderIndex {
+                            ui: chalk_ir::UniverseIndex::ROOT,
+                            idx: 0,
+                        }),
+                    ),
+                    value: chalk_ir::ConstValue::InferenceVar(chalk_ir::InferenceVar::from(0)),
```
[the remark this answers is not in the document]
Comment (jackh726) [reply tests/test/type_flags.rs]: oh this is const, nvm

**Comment by jackh726** [comment]
@matthewjasper can you take another look at this? I've rebased and updated per your review.

**Review by matthewjasper — COMMENTED** [review]
r=me with comment addressed

**Pull request merged**